### ISSUE TITLE
Some issues about CJK text recognition

### ISSUE THREAD

**Issue by gexgd0419** [TheJoeFin/Text-Grab#191]
Chinese and Japanese text don't have spaces between characters, but Windows OCR engine recognizes each "character" as a "word" and puts spaces between each character. This has been fixed by combining the content of each word manually to get rid of the spaces when the language is CJK. However, Korean text actually uses spaces between words, and Windows OCR engine recognizes the words correctly. The spaces in Korean text should be kept.

![image](https://user-images.githubusercontent.com/55008943/192427503-cdac7507-c320-4dd5-83fa-9da667fc8dcb.png)

Chinese and Japanese OCR engine can recognize English words as well. So, when the content has both Chinese/Japanese text and English words, the spaces between English words are gone. We could check each "word" to see if it's an English word, and then put a space between two English words.

**Comment by TheJoeFin** [comment]
Thank you for highlighting the differences in how Korean treats word spaces vs Chinese and Japanese. This should be an easy fix. As for OCR different languages at the same time, this is currently not possible with the current OCR API used by Text Grab. The language does not get detected on a word by word basis. If that was the case I would absolutely look for each words language and remove spaces if needed.

Thank you for you feedback, and if you could share some reference images along with the desired output as text, I would appreciate it greatly!

**Comment by TheJoeFin** [comment]
I'm glad to return to this issue and say there is a solution! I will be doing some testing soon mostly based upon the work of https://github.com/maggch97 in the PowerToys Repo specifically this PR https://github.com/microsoft/PowerToys/pull/20926

Hopefully this can be resolved soon and this method will open the door for more accurate OCR results in all languages.

**Comment by gexgd0419** [comment]
`\p{IsCJKUnifiedIdeographs}` matches only ideographs, or what's called "Chinese character/Hanzi" in Chinese and "kanji" in Japanese.

CJK punctuation marks (`IsCJKSymbolsandPunctuation`), Japanese hiragana (`IsHiragana`) and katakana (`IsKatakana`) are not in the range of "ideographs". But those are also recognized as single-character "words", and they should not be surrounded with spaces either.

(The Korean alphabet, Hangul, also has its own range, `IsHangulSyllables`. But Korean text uses spaces to separate words, and Windows OCR engine can do this well, so this is not needed)

I'm not sure what other kinds of characters/symbols should be added to the list, though. There are lots of other characters and symbols that fall in different Unicode ranges.

My initial thought about this issue was, we could check each one-character "word" (because multiple-character words cannot be CJK characters) to see if it matches `[A-Za-z0-9]` and treat all other characters as if they were ~~CJK~~ "no spaces needed" characters.

**Comment by TheJoeFin** [comment]
@gexgd0419 thank you for highlighting the different ranges and if they need or do not need spaces. Since `IsCJKUnifiedIdeographs` matches too broadly what do you think of this approach:

```
    public static bool IsSpaceJoiningLanguage(this string wordToBeChecked)
    {
        Regex broadCJK = new(@"\p{IsCJKUnifiedIdeographs}|\p{IsHiragana}|\p{IsKatakana}");
        Match cjkMatch = broadCJK.Match(wordToBeChecked);

        if (cjkMatch.Success)
        {
            Regex checkForKorean = new(@"\p{IsHangulSyllables}");
            Match koreanMatch = checkForKorean.Match(wordToBeChecked);
            if (koreanMatch.Success)
                return true;

            return false;
        }

        return true;
    }
```

**Comment by gexgd0419** [comment]
## The OCR language affects word division

I think that how to divide characters into "words" is decided by the OCR engine of that language.

For example, I tried to OCR the following image with Chinese, Japanese and Korean OCR language. (These are all Chinese characters, but I put two spaces in them)
![image](https://user-images.githubusercontent.com/55008943/194481216-269f7e93-06d9-4bd6-93d8-dee7f4298ff7.png)

Results:
Chinese & Japanese: 12 words, `文 字 文 字 文 字 文 字 文 字 文 字`
Korean: 3 words, `文字文字 文字文字文 字文字`

They all recognized the correct characters, but only Korean OCR engine divided the characters into words by the spaces I put between characters.

(Historically, Koreans have used Chinese characters in their writing system, and the adopted Chinese characters are called "hanja". Now they mainly use their own alphabet, "hangul". But Windows Korean IME and OCR still support hanja.)

So, I think we should still check the OCR language specified by the user, instead of guessing whether its language needs spaces purely from the recognition result. If the OCR language is Korean, the recognition result should be treated like an English recognition result, and we don't need to check for Korean characters separately.

## The CJK Unified Ideographs range

`IsCJKUnifiedIdeographs` (code point range `4E00 - 9FFF`), containing over 20,000 characters, is a broad range.

But this range only contains "ideographs" (Chinese characters/kanjis/hanjas). And that's not all of them.

Many Unicode blocks have names like "CJK compatibility ideographs", "CJK compatibility forms", "CJK unified ideographs extension A", "CJK radical supplements", etc., which added another few thousand characters to Unicode. There are so many CJK ideographs that Unicode added many "supplement" and "extension" blocks.

However, `IsCJKUnifiedIdeographs` still has almost all of the frequently-used characters. Although there are tens of thousands of Chinese characters in total, just a few thousand are still frequently used today. If you pick a random Chinese character outside the range of `IsCJKUnifiedIdeographs` and pass it to OCR engine, there's a high chance that it cannot be recognized correctly. So it's possible that the OCR engine can only recognize (a subset of) characters in this range.

## Punctuation marks

Chinese and Japanese texts usually don't need spaces around punctuations.

Some punctuations are in `IsGeneralPunctuation`. CJK-exclusive punctuations are in `IsCJKSymbolsandPunctuation`. Punctuations in the Basic Latin range (the standard ASCII range, 0 - 127), and their full-width forms in `IsHalfwidthandFullwidthForms`, can also be used.

For example,
- `“” … ※` are in `IsGeneralPunctuation`.
- `《》【】「」、。` are in `IsCJKSymbolsandPunctuation`.
- `，？！：；（）` are in `IsHalfwidthandFullwidthForms`, because they are the full width forms of these ASCII characters: `,?!:;()`
- ASCII characters such as `@#%&*` can also be used.

Note that `IsHalfwidthandFullwidthForms` has the full-width forms of other ASCII characters as well, including numbers and English letters.
Compare: `Abcd123` (half-width, ASCII) and `Ａｂｃｄ１２３` (full-width, in `IsHalfwidthandFullwidthForms`).
But it seems that there won't be full-width numbers/letters in the OCR result, so maybe it's OK to treat all characters in `IsHalfwidthandFullwidthForms` as CJK punctuations.

## Other symbols

There are many other Unicode blocks that contain symbols instead of the script of some language. For example, `IsArrows` only has arrow symbols like `←↑→↓`.

I guess that most of those symbols need no spaces around them.

However, it seems that the OCR engine cannot recognize many of those symbols, either.

## So ... what should we do?

I think at least `IsCJKUnifiedIdeographs`, `IsHiragana`, `IsKatakana`, `IsGeneralPunctuation`, `IsCJKSymbolsandPunctuation`, and the punctuation marks in `IsHalfwidthandFullwidthForms` should be treated as "no space needed" characters. This is not a complete set, but it should be enough for most CJK contents.

As for Korean, just let the OCR engine handle it. No need to "check for CJK then check for Korean".

Or, maybe we can check for characters that are **NOT** CJK characters. Not sure which way is easier, though.

**Comment by SodaWithoutSparkles** [comment]
I am from the powertoys issue above, and I know nothing about windows/C# programming, but I do have experience programming in bash, C, and python. I am afraid I know nothing about the `\p{IsCJK...}` mentioned above, but I have thought of a few points to test:

Some punctuations have chinese and english varients, the following is a list of commonly used punctuations and their chinese variants, separated by a `|`. Chinese varients are placed first

- （）|()
- ，|,
- 。|.
- ？|?
- ！|!
- ：|:
- ；|;
- ［］|[]

Another point to note is that when typing Chinese you have the option to press `SHIFT` to switch between Chinese and English. This does not work when using text extractor. It still removes all the spaces between english words. 

Here is another sample piece of text with mixed language: 

快捷鍵指南 (shortcut menu) 是一個工具讓你可以在忘記快捷鍵（人是很健忘的）時調用以獲取提示 (to get tips when you forgot shortcuts)。它具有方便、快捷、易用的特性，能夠讓你儘快記住快捷鍵並以更高效率操作。有用嗎？當然有！有研究表明這能有69%的效率提昇 (Wow! A 69% increase?)。

有部分程序員在忘記轉換輸入法的時候把英文分號 ( ; ) 輸入成了中文分號 (；)，這會導致無法 compile，導致出錯。在push之前記得check下你com唔compile到。

**Comment by gexgd0419** [comment]
> 快捷鍵指南 (shortcut menu) 是一個工具讓你可以在忘記快捷鍵（人是很健忘的）時調用以獲取提示 (to get tips when you forgot shortcuts)。它具有方便、快捷、易用的特性，能夠讓你儘快記住快捷鍵並以更高效率操作。有用嗎？當然有！有研究表明這能有69%的效率提昇 (Wow! A 69% increase?)。
> 
> 有部分程序員在忘記轉換輸入法的時候把英文分號 ( ; ) 輸入成了中文分號 (；)，這會導致無法 compile，導致出錯。在push之前記得check下你com唔compile到。

I took a screenshot of the text above and tried to OCR it using Windows OCR engine.

Here's the result using Simplified Chinese as the OCR language (the spaces are not removed):
> 快 捷 鍵 指 南 (shortcut menu) 是 一 個 工 具 讓 你 可 以 在 忘 記 快 捷 鍵 （ 人 是 很 健 忘 的 ） 時 調 用 以 獲 取 提 示 ()o g et tips when you forgot shortcuts)o 它 具 有 方 便 、 快 捷 、 易 用 的 特 性 ， 能 麴 讓 你 儘 快 記 住 快 捷 鍵 並 以 更 高 效 率 操 作 。 有 用 嗎 ？ 當 然 有 ！ 有 研 究 表 明 這 能 有 69 ％ 的 效 率 提 舁 (Wow ！ A 69 ％ increase?)o 有 部 分 程 序 員 在 忘 記 轉 換 輸 入 法 的 時 候 把 英 文 分 號 （ ； ） 輸 入 成 了 中 文 分 號 (; ), 這 會 導 致 無 法 compile, 導 致 出 錯 。 在 push 之 前 記 得 check 下 你 comßäcompile 到 。

And here's the result using Traditional Chinese as the OCR language (the spaces are not removed):
> 快 捷 鍵 指 南 (shortcut menu) 是 一 個 工 具 讓 你 可 以 在 忘 記 快 捷 鍵 ( 人 是 很 健 忘 的 ) 時 調 用 以 獲 取 提 示 ()o get tips when you forgot shortcuts)o 它 具 有 方 便 、 快 捷 、 易 用 的 特 性 , 能 夠 讓 你 儘 快 記 住 快 捷 鍵 並 以 更 高 效 率 操 作 。 有 用 嗎 ? 當 然 有 ! 有 研 究 表 明 這 能 有 69 % 的 效 率 提 昇 (Wow! A 69 % increase?)o 有 部 分 程 序 員 在 忘 記 轉 換 輸 入 法 的 時 候 把 英 文 分 號 ( ; ) 輸 入 成 了 中 文 分 號 ( ; ) , 這 會 導 致 無 法 compile , 導 致 出 錯 。 在 push 之 前 記 得 check 下 你 com 唔 compile 到 。

Notice those punctuation marks.
Seems that for a punctuation character that has both a half-width variant (ASCII or "English" character) and a full-width variant (more used in Chinese text), Simplified Chinese OCR prefers full-width (unless when it's near an English word), and Traditional Chinese OCR prefers half-width. This would cause some problems, as English text also uses half-width punctuations.

I think that the word's length can be used to distinguish a punctuation's language context.

In English, punctuations are attached to a word.
For example, the OCR result of `Yeah, right.` consists of two words: `Yeah,` and `right.`. Both words contain the punctuation mark following them.

Chinese and Japanese characters, including punctuations, are separated into single-character words.
(Then the OCR engine tries to combine the words with a space between each word, that's why Chinese and Japanese OCR results are full of spaces)
For example, the OCR result of `你好，` consists of three words: `你`, `好`, and `，`, where the punctuation `，` becomes a single-character word.

**Comment by TheJoeFin** [comment]
<img width="1000" alt="GitHub-Comment-Mixed-Lang-Text" src="https://user-images.githubusercontent.com/7809853/195738690-730a8efc-d06a-48f5-bd0d-e8d5a4796232.png">

I have updated the logic in Text Grab and now the output of the OCR from the above image when using zh-Hans-CN is:

> 快捷鍵指南 (shortcut menu) 是一個工具讓你可以在忘記快捷鍵（人是很健忘的）時調用以獲取提示 (to get tips when you forgot shortcuts)。它具有方便、快捷、易用的特性，能夠讓你儘快記住快捷鍵並以更高效率操作。有用嗎？當然有！有研究表明這能有69%的效率提昇 (Wow! A 69% increase?)。
> 
> 有部分程序員在忘記轉換輸入法的時候把英文分號 ( ; ) 輸入成了中文分號 (；)，這會導致無法 compile，導致出錯。在push之前記得check下你com唔compile到。

Not perfect, but much better results in my opinion. Does anyone have any feedback on this update?

**Comment by SodaWithoutSparkles** [comment]
> Not perfect, but much better results in my opinion. Does anyone have any feedback on this update? 

Way better. There are some imperfections but its acceptable.

**Comment by gexgd0419** [comment]
> <img alt="GitHub-Comment-Mixed-Lang-Text" width="1000" src="https://user-images.githubusercontent.com/7809853/195738690-730a8efc-d06a-48f5-bd0d-e8d5a4796232.png">
> 
> I have updated the logic in Text Grab and now the output of the OCR from the above image when using zh-Hans-CN is:
> 
> > 快捷鍵指南 (shortcut menu) 是一個工具讓你可以在忘記快捷鍵（人是很健忘的）時調用以獲取提示 (to get tips when you forgot shortcuts)。它具有方便、快捷、易用的特性，能夠讓你儘快記住快捷鍵並以更高效率操作。有用嗎？當然有！有研究表明這能有69%的效率提昇 (Wow! A 69% increase?)。
> >
> > 有部分程序員在忘記轉換輸入法的時候把英文分號 ( ; ) 輸入成了中文分號 (；)，這會導致無法 compile，導致出錯。在push之前記得check下你com唔compile到。
> 
> Not perfect, but much better results in my opinion. Does anyone have any feedback on this update?

The "output of the OCR" shown above is EXACTLY the same as the source text.

Is it possible that the source text, instead of the real OCR result, was mistakenly pasted here?

**Comment by TheJoeFin** [comment]
@gexgd0419 good catch! even the line returns are matched up, which doesn't happen with the OCR. Here is the OCR Result

> 快捷鍵指南 (shortcut menu)是一個工具讓你可以在忘記快捷鍵 （人是很健忘的 ）時調用
以獲取提示 ()o get tips when you fo rgot shortcuts)o它具有方便 、快捷 、易用的特性 ，能
麴讓你儘快記住快捷鍵並以更高效率操作 。有用嗎 ？當然有 ！有研究表明這能有 69 ％的效
率提舁 (Wow! A 6900 increase?)o
有部分程序員在忘記轉換輸入法的時候扌巴英文分號 （ ； ）輸入成了中文分號 （ ； ） ，這會導致
無氵去 compile,導致出錯 。在 push之前旨己得 check下你 com唔 compile到 。

Not perfect, but I think it is an improvement.

**Comment by SodaWithoutSparkles** [comment]
still some words are spilt. Its acceptable tho. researches shows that you can swap the order of words of chinese and you can still read it. words spiltting looks strange but we can understand

**Comment by gexgd0419** [comment]
I tried the version of Text Grab at this commit: ec77dbaf5919e57ed3a258c3898fa7695590ea82. Here's what I think could still be improved.

First, I noticed this line:

https://github.com/TheJoeFin/Text-Grab/blob/ec77dbaf5919e57ed3a258c3898fa7695590ea82/Text-Grab/Utilities/StringMethods.cs#L345

where the regex is:

`\p{IsCJKUnifiedIdeographs}|{IsCJKSymbolsandPunctuation}|{IsHiragana}|{IsKatakana}|{IsHalfwidthandFullwidthForms}`

It didn't work on my system and only `IsCJKUnifiedIdeographs` was matched, so I changed the regex to:

`\p{IsCJKUnifiedIdeographs}|\p{IsGeneralPunctuation}|\p{IsCJKSymbolsandPunctuation}|\p{IsHiragana}|\p{IsKatakana}|\p{IsHalfwidthandFullwidthForms}`

where I added `IsGeneralPunctuation` to include punctuations such as `“” … ※`.

And here's the logic to join the words:

https://github.com/TheJoeFin/Text-Grab/blob/ec77dbaf5919e57ed3a258c3898fa7695590ea82/Text-Grab/Utilities/ImageMethods.cs#L189-L200

which means that there's always a space before an English word, but whether there is a space after it or not is determined by the word after it. If an English word is put between Chinese characters, there will be a space before the word, but there won't be any spaces after the word, because the "word" after it is not in a "SpaceJoiningLanguage".

When a word in a "SpaceJoiningLanguage" meets a word not in a "SpaceJoiningLanguage", the space between them is optional, and it can even be a user-adjustable option. (Microsoft Word lets user choose whether to show a space between Chinese text and English words/numbers, for example)

But the behavior should be consistent. `出bug了` and `出 bug 了` are acceptable, but `出 bug了` is not.

The program should keep track of whether the previous word is in a "SpaceJoiningLanguage", and compare it with the current word. Only when both words are not in "SpaceJoiningLanguage" should the space between them be removed, if you choose to add a space between Chinese and English words.

Also, I noticed that in previous versions Text Grab would check `selectedLanguage.LanguageTag` to determine the language, but now it just put every word through a regex regardless of the OCR language. Only the OCR languages that starts with `zh-` or is `ja` need special treatment.

A word containing Chinese characters doesn't mean that the word is in Chinese (or Japanese). Technically, Korean text can also have Chinese characters, and the Korean OCR engine can recognize them, although Chinese characters are not used often. If a Korean word contains Chinese characters, this word will match that regex, but we know that Korean text needs spaces. Even a Chinese-character-only word can be a Chinese word, a Japanese word, and a Korean word at the same time, so it should be interpreted based on the user's selected OCR language.

An English punctuation (half-width) can be used in a non-English context as well. Traditional Chinese (zh-Hant) OCR engine seems to prefer English punctuations.

---

Anyway, here's my PR #197 that tries to fix this issue. It changed from matching Chinese/Japanese characters to matching non-Chinese/Japanese letters so that all symbols are joined without spaces.

Now the Simplified Chinese OCR result of the text above is like this:

> 快捷鍵指南 (shortcut menu) 是一個工具讓你可以在忘記快捷鍵（人是很健忘的）時調用
> 以獲取提示 ()o get tips when you forgot shortcuts)o 它具有方便、快捷、易用的特性，能
> 麴讓你儘快記住快捷鍵並以更高效率操作。有用嗎？當然有！有研究表明這能有 69 ％的效
> 率提舁 (Wow ！ A 69 ％ increase?)o
> 有部分程序員在忘記轉換輸入法的時候把英文分號（；）輸入成了中文分號（；），這會導致
> 無法 compile, 導致出錯。在 push 之前記得 check 下你 com 唔 compile 到。

And the Traditional Chinese OCR result:

> 快捷鍵指南 (shortcut menu) 是 H 固工具讓你可以在忘記快捷鍵(人是很健忘的)時調用
> 以獲取提示 get tips when you forgot shortcuts)o 它具有方便、快捷、易用的特性,能
> 夠讓你儘快記住快捷鍵並以更高效率操作。有用嗎?當然有!有研究表明這能有 69 %的效
> 率提昇 (wow! A 69 % increase?)o
> 有部分程序員在忘記轉換輸入法的時候把英文分號(;)輸入成了中文分號(;),這會導致
> 無法 compile, 導致出錯。在 push 之刖記得 check 下你 comßäcompile 到。
